confdb/confdb_test.go

```diff
@@ -3645,3 +3647,125 @@ func (*viewSuite) TestUnsetNestedList(c *C) {
+var subkeyOnly = regexp.MustCompile(`\[[^.\[]*|[^.\[]+`)
+
+func (*viewSuite) TestSubkeyDivideRegex(c *C) {
+	s := "[∀.∃*-l}].[[.."
+	d := subkeyDivide.FindAllString(s, -1)
+	c.Check(len(d), Equals, 8)
+	c.Check(d[0], Equals, "[∀")
+	c.Check(d[1], Equals, ".")
+	c.Check(d[2], Equals, "∃*-l}]")
+	c.Check(d[3], Equals, ".")
+	c.Check(d[4], Equals, "[")
+	c.Check(d[5], Equals, "[")
+	c.Check(d[6], Equals, ".")
+	c.Check(d[7], Equals, ".")
+}
+
+func hasValidSubkeys(s string, o confdb.ParseOptions) bool {
+	subStrings := subkeyOnly.FindAllString(s, -1)
+	for _, ss := range subStrings {
+		// If the substring doesn't match the generic case
+		if !confdb.ValidSubkey.MatchString(ss) &&
+			// if it doesn't match the placeholder case when placeholders are allowed
+			!(o.AllowPlaceholders && confdb.ValidPlaceholder.MatchString(ss)) &&
+			// if it doesn't match the index case when partial paths are allowed and indices are not forbidden
+			!(!o.ForbidIndexes && o.AllowPartialPath && confdb.ValidIndexSubkey.MatchString(ss)) &&
+			// if it doesn't match index placeholders when placeholders and partial paths are allowed
+			!(o.AllowPlaceholders && o.AllowPartialPath && confdb.ValidIndexPlaceholder.MatchString(ss)) {
+			return false
+		}
+	}
+	return true
+}
+
+func hasEmptySubkey(s string, o confdb.ParseOptions) bool {
+	subStrings := subkeyDivide.FindAllString(s, -1)
+	if len(subStrings) == 0 {
+		return true
+	}
+	if subStrings[0] == "." || subStrings[len(subStrings)-1] == "." {
+		return true
+	}
+	// Only if partial paths are allowed can a path begin with '['
+	if strings.HasPrefix(subStrings[0], "[") && !o.AllowPartialPath {
+		return true
+	}
+	for i, ss := range subStrings {
+		// A subkey can never begin with a '[' unless it's the first rune
+		// Two successive '.' means an empty string subkey
+		if i > 0 && subStrings[i-1] == "." && (strings.HasPrefix(ss, "[") || ss == ".") {
+			return true
+		}
+	}
+	return false
+}
+
+func fuzzHelper(f *testing.F, o confdb.ParseOptions, seed string) {
+	wrapper := func(s string) ([]confdb.Accessor, error) {
+		return confdb.ParsePathIntoAccessors(s, o)
+	}
+	f.Add(seed)
+	f.Fuzz(func(t *testing.T, s string) {
+		accessors, err := wrapper(s)
+		if err != nil && strings.Contains(err.Error(), "invalid subkey") && !hasValidSubkeys(s, o) {
+			t.Skip()
+		}
+		if err != nil && err.Error() == "cannot have empty subkeys" && hasEmptySubkey(s, o) {
+			t.Skip()
+		}
+		if err != nil {
+			t.Errorf("encountered error %s with input %s", err, s)
+		}
+		expected := subkeyOnly.FindAllString(s, -1)
+		if len(accessors) != len(expected) {
+			t.Errorf("unexpected number of accessors %d vs. %d", len(accessors), len(expected))
+		}
+		for i, e := range expected {
+			if strings.HasPrefix(e, "[{") {
+				if accessors[i].Type() != confdb.IndexPlaceholderType {
+					t.Errorf("unexpected type of accessor %v with name %s for element %s", accessors[i].Type(), accessors[i].Name(), e)
```

Comment:
The Type() is backed by a uint and it doesn't implement stringer so this will look like `unexpected type of accessor 2 with name...`. Maybe we can add a String() method to AccessorType